error_handling/README.md

```diff
@@ -0,0 +1,223 @@
+that will always fail:
+
+```go
+func errorFunc(inputPtr, secretPtr uint64) uint64 {
+	err := errors.New("expected error")
+	return writeError(err)
+}
+```
+
+Both of these functions will successfully complete their execution on the 
+Blocky AS server and produce an attestation over their output. What we'd like
+is an easy way to tell whether the function succeeded or failed due to a runtime
+error.
+
+### Step 2: Use the result pattern
+
+To do this, we can use the [result pattern](https://en.wikipedia.org/wiki/Result_type).
+In [`main.go`](./main.go), we define a `Result` struct that can hold either
+a successful result or an error:
+
+```go
+type Result struct {
+	Success bool
+	Error   string
+	Value   any
+}
+```
+
+The `Success` field indicates whether the function succeeded or failed. The
+`Error` field holds an error message if the function failed (`Success` is
+`false`). If the function succeeds (`Success` is `true`), the `Error` field
+should be disregarded. The `Value` field holds the result of the function if it
+succeeded (`Success` is `true`). If the function fails (`Success` is `false`),
+the `Value` field should be disregarded. Note, that this three-field struct
+design of `Result` is a slight departure from the traditional result pattern, 
+where typically a `Value` field of type `any` might hold either an error message, or another return
+type. Having the three fields, however, allows for one pass parsing of `Result`
+structs in the client code.
+
+To return a `Result` to user, we need to serialize to bytes and send them to
+the `as.WriteToHost` function. Let's say that we want to use JSON to serialize
+the `Result` struct. 
+
+We can define the `writeOutput` function to take in our function output, as 
+`any`, put it in a `Result` struct, serialize it, and send it to 
+`as.WriteToHost`:
+
+```go
+func writeOutput(output any) uint64 {
+	result := Result{
+		Success: true,
+		Value:   output,
+	}
+	data, err := json.Marshal(result)
+	if err != nil {
+		as.Log(fmt.Sprintf("Error marshalling result: %s", err))
+		return writeError(err)
+	}
+	return as.WriteToHost(data)
+}
+```
+
+As you see, we have a challenge here in that the `json.Marshal` function itself
+can fail. In this case, we can use the `writeError` function to report that 
+error. But wouldn't we run into the same, chicken-and-egg problem if we
+encountered an error in the `writeError` function? Let's take a look.
+
+Our `writeError` function is defined as:
+
+```go
+func writeError(err error) uint64 {
+	data := Result{}.jsonMarshalWithError(err)
+	return as.WriteToHost(data)
+}
+```
+
+and uses the receiver function `jsonMarshalWithError`:
+
+```go
+func (r Result) jsonMarshalWithError(err error) []byte {
+	if err == nil {
+		err = errors.New("jsonMarshalWithError invoked with nil error")
+	}
+	resultStr := fmt.Sprintf(
+		`{ "Success": false, "Error": "%v" , "Value": null }`,
+		err,
+	)
```

Comment:
I wonder if it would be better to use
```
resultStr := fmt.Sprintf(
    `{ "Success": false, "Error": "%s", "Value: null }`, 
    err.Error(),
)
```

The %v verb is go specific as where %s is used in most Sprintf type of environments, so more developers will know it already.

Reply:
Sure, good suggestion. Now that we are safe to use err.Error it will be cleaner